crypto/x509/utils/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+        if self.offset == self.pem_slice.len() {
+            None
+        } else {
+            X509Certificate::from_der(&self.pem_slice[self.offset].contents)
+                .map(|(_, x509)| {
+                    self.offset += 1;
+                    x509
+                })
+                .ok()
+        }
+    }
+}
+
+/// A trait used to monkey-patch an X509Certificate parsing iterator over a
+/// slice of [`Pem`] objects.
+pub trait X509CertificateIterable {
+    fn iter_x509(&self) -> X509CertificateIter;
+}
+
+/// Anything which can be referenced as a Pem slice gets a method to iterate
+/// certificates, and verify it is a well-formed single-path certificate chain.
+impl<T: AsRef<[Pem]>> X509CertificateIterable for T {
+    fn iter_x509(&self) -> X509CertificateIter {
+        X509CertificateIter::new(self)
+    }
+}
+
+/// A trait used to monkey-patch an X509Certificate chain verifier onto a slice
+/// of X509Certificate objects.
+pub trait X509CertificateChain {
+    /// Verify the chain (checks validity, signatures, and CA extension of each
+    /// element)
+    fn verify_chain(&self) -> bool;
+}
+
+impl<'a, T: AsRef<[X509Certificate<'a>]>> X509CertificateChain for T {
+    fn verify_chain(&self) -> bool {
+        let mut previous = None;
+        let mut cert_count = 0usize;
+
+        if self.as_ref().is_empty() {
+            return false;
+        }
+
+        for (index, cert) in self.as_ref().iter().enumerate() {
+            // If the cert wasn't signed by the preceeding cert (or itself, if first), fail.
+            if cert.verify_signature(previous).is_err() {
+                return false;
+            }
+
+            // If the cert isn't valid (temporally), fail.
+            if !cert.validity().is_valid() {
+                return false;
+            }
+
+            // Update state for the next iteration
+            previous = Some(&cert.tbs_certificate.subject_pki);
+            // Update the number of certificates which have passed validation
+            cert_count = index + 1;
+
+            if !cert.tbs_certificate.is_ca() {
+                break;
+            }
+        }
+
+        // If any of the certs didn't pass verification, or there was a non-CA
+        // cert in the middle of the chain, fail.
+        if cert_count != self.as_ref().len() {
+            return false;
+        }
+
+        // If the last cert in the chain is a CA, fail.
+        !self.as_ref()[cmp::max(0, cert_count - 1)]
+            .tbs_certificate
+            .is_ca()
+    }
+}
+
+/// A list of key types supported by both X.509 and mc-crypto-keys.
+pub enum PublicKeyType {
+    /// The public key type is invalid
+    Invalid,
+    Unknown(Vec<u8>),
+    Ed25519(Ed25519Public),
+}
+
+/// Monkey-patch support for extracting mc-crypto-keys public keys from X509
+/// certificates.
+pub trait X509KeyExtrator {
+    /// Try to retrieve the public key.
+    fn mc_public_key(&self) -> PublicKeyType;
```

Comment:
It might be better to return `Result<PublicKeyType>` and treat the Invalid case as an error?

Reply:
I'm also in favor of that - any reason you opted for an Enum variant for this?

Reply:
Simply not thinking clearly, this has been fixed. ;-)